File: algorithms/divide-and-conquer/karatsuba-multiplication/swift/Karatsuba.swift
import Foundation

func karatsuba(_ arr: [Int]) -> Int {
    func multiply(_ x: Int, _ y: Int) -> Int {
        if x < 10 || y < 10 { return x * y }

        let nx = String(abs(x)).count
        let ny = String(abs(y)).count
        let n = max(nx, ny)
        let half = n / 2
        var power = 1
        for _ in 0..<half { power *= 10 }

        let x1 = x / power, x0 = x % power
        let y1 = y / power, y0 = y % power

        let z0 = multiply(x0, y0)
        let z2 = multiply(x1, y1)
        let z1 = multiply(x0 + x1, y0 + y1) - z0 - z2

        return z2 * power * power + z1 * power + z0
    }

    return multiply(arr[0], arr[1])
}

File: algorithms/divide-and-conquer/karatsuba-multiplication/rust/karatsuba.rs
fn num_digits(n: i64) -> u32 {
    if n == 0 { return 1; }
    let mut count = 0;
    let mut val = n.abs();
    while val > 0 {
        count += 1;
        val /= 10;
    }
    count
}

fn multiply(x: i64, y: i64) -> i64 {
    if x < 10 || y < 10 {
        return x * y;
    }

    let n = num_digits(x).max(num_digits(y));
    let half = n / 2;
    let power = 10i64.pow(half);

    let (x1, x0) = (x / power, x % power);
    let (y1, y0) = (y / power, y % power);

    let z0 = multiply(x0, y0);
    let z2 = multiply(x1, y1);
    let z1 = multiply(x0 + x1, y0 + y1) - z0 - z2;

    z2 * power * power + z1 * power + z0
}

pub fn karatsuba(arr: &[i32]) -> i32 {
    multiply(arr[0] as i64, arr[1] as i64) as i32
}
